fix(ivy): ensure component/directive `class` selectors are properly understood

matsko opened the pull request:
Angular allows for `<ng-content>` elements to include a selector which
filters which content-projected entries are inserted into the container
depending on whether or not the selector is matched.

With Ivy this feature has not fully worked due to the massive changes
that took place inside of Ivy's styling algorithm code (which is
responsible for assigning classes and styles to an element). This
fix ensures that content-projection can correctly identify which slot
an element should be placed into when class-based selectors are used.

kara reviewed (changes requested):
Can you remove the fixmes for FW-833 (or replace desc with new root cause if there are additional issues)?

mary-poppins commented:
You can preview 2277521 at https://pr27849-2277521.ngbuilds.io/.

kara reviewed (approved):
~~LGTM~~

kara reviewed (changes requested):
@matsko Actually it looks like the tests added (for FW-833) are failing on CI. Can you take a look?

matsko committed:
fix(ivy): ensure component/directive `class` selectors are properly understood

Angular allows for `<ng-content>` elements to include a selector which
filters which content-projected entries are inserted into the container
depending on whether or not the selector is matched.

With Ivy this feature has not fully worked due to the massive changes
that took place inside of Ivy's styling algorithm code (which is
responsible for assigning classes and styles to an element). This
fix ensures that content-projection can correctly identify which slot
an element should be placed into when class-based selectors are used.

pkozlowski-opensource commented on packages/core/test/linker/projection_integration_spec.ts:
```diff
@@ -81,22 +81,35 @@ describe('projection', () => {
-          }
-        });
-        const main = TestBed.createComponent(MainComp);
+  fit('should project a single class-based tag', () => {
```
`fit` left

mary-poppins commented:
You can preview 552b267 at https://pr27849-552b267.ngbuilds.io/.

mary-poppins commented:
You can preview 386a663 at https://pr27849-386a663.ngbuilds.io/.

mary-poppins commented:
You can preview cf66990 at https://pr27849-cf66990.ngbuilds.io/.

kara reviewed (approved):
LGTM

matsko commented:
https://tap.corp.google.com/ui#id=OCL:228943859:BASE:228949472:1547246925594:82175f07

mary-poppins commented:
You can preview 34979bd at https://pr27849-34979bd.ngbuilds.io/.

matsko commented:
http://test/OCL:228943859:BASE:229271551:1547509864607:f18a0f5

mary-poppins commented:
You can preview cc14acc at https://pr27849-cc14acc.ngbuilds.io/.
